hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RSRpcServices.java

```diff
@@ -3762,6 +3766,40 @@ public ClearRegionBlockCacheResponse clearRegionBlockCache(RpcController control
     return builder.setStats(ProtobufUtil.toCacheEvictionStats(stats.build())).build();
   }
 
+
+  @Override
+  public CompleteCompactionResponse completeCompaction(RpcController controller,
+      CompleteCompactionRequest request) throws ServiceException {
+    RegionInfo regionInfo = ProtobufUtil.toRegionInfo(request.getRegionInfo());
+    ColumnFamilyDescriptor family = ProtobufUtil.toColumnFamilyDescriptor(request.getFamily());
+    LOG.debug("Region server receive complete compaction for region: {}, cf: {}",
+      regionInfo.getRegionNameAsString(), family.getNameAsString());
+    boolean success = false;
+    HRegion onlineRegion = regionServer.getOnlineRegion(regionInfo.getRegionName());
+    if (onlineRegion != null) {
+      HStore store = onlineRegion.getStore(family.getName());
+      if (store != null) {
+        if (store.getForceMajor()) {
+          store.setForceMajor(request.getNewForceMajor());
+        }
+        List<String> selectedFiles =
+            request.getSelectedFilesList().stream().collect(Collectors.toList());
+        List<String> newFiles = request.getNewFilesList().stream().collect(Collectors.toList());
+        try {
+          success = store.completeCompaction(null, selectedFiles, null, newFiles);
```

Comment:
If we could write HFile directly into the data directory, here the completion will be easier. Just a note.

Reply:
Recorded in TODO